### PR TITLE
Fix1444

### LINKED ISSUE #1444: Design and implement IPackageManagerProvider API

### PR DESCRIPTION
**Pull request by zhili1208**
fixed https://github.com/NuGet/Home/issues/1444

added API for integrating other package manager

@yishaigalatzer @emgarten @deepakaravindr @feiling @danliu @MeniZalzman

### CODE REVIEW

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+﻿using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.VisualStudio
+{
+    /// <summary>
+    /// Interface for integrating other package manager
```
Comment (emgarten) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: Provide a longer description here, comments on public VS interfaces will become the documentation.

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        /// </summary>
+        string Description { get; }
+
+        /// <summary>
+        /// determine do we need to direct to integrated package manager from NuGet UI for specified package and project.
+        /// </summary>
+        /// <param name="packageId">Current package id</param>
+        /// <param name="projectName">Unique project name</param>
+        /// <param name="token">Cancellation Token</param>
+        /// <returns></returns>
```
Comment (emgarten) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: Fill out the return value

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        /// </summary>
+        string PackageManagerID { get; }
+
+        /// <summary>
+        /// Description for package manager
+        /// </summary>
+        string Description { get; }
+
+        /// <summary>
+        /// determine do we need to direct to integrated package manager from NuGet UI for specified package and project.
```
Comment (emgarten) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: Capitalize the first letter on this one and the next summary.

File: src/NuGet.Clients/PackageManagement.UI/UserInterfaceService/NuGetUIContext.cs [hunk 4]
```diff
@@ -14,6 +15,7 @@ namespace NuGet.PackageManagement.UI
     public abstract class NuGetUIContextBase : INuGetUIContext
     {
         private readonly NuGetProject[] _projects;
+        private readonly IEnumerable<IPackageManagerProvider> _packageManagerProviders;
```
Comment (emgarten) [inline src/NuGet.Clients/PackageManagement.UI/UserInterfaceService/NuGetUIContext.cs]: This isn't used anywhere, should it be removed from this class?

File: src/NuGet.Clients/VsExtension/VisualStudioUIContextFactory.cs [hunk 5]
```diff
@@ -22,6 +24,9 @@ internal class VisualStudioUIContextFactory : INuGetUIContextFactory
         private readonly IOptionsPageActivator _optionsPage;
         private readonly ISettings _settings;
         private readonly IDeleteOnRestartManager _deleteOnRestartManager;
+        private readonly IList<IPackageManagerProvider> _packageManagerProviders;
+        // only pick up at most three integrated package managers
+        private readonly int _maxPackageManager = 3;
```
Comment (emgarten) [inline src/NuGet.Clients/VsExtension/VisualStudioUIContextFactory.cs]: This can be a const since it is private.

File: src/NuGet.Clients/VsExtension/VisualStudioUIContextFactory.cs [hunk 6]
```diff
@@ -62,7 +69,25 @@ public INuGetUIContext Create(NuGetPackage package, IEnumerable<NuGetProject> pr
                 actionEngine,
                 _restoreManager,
                 _optionsPage,
-                projects);
+                projects,
+                _packageManagerProviders);
+        }
+
+        private List<IPackageManagerProvider> Sort(IEnumerable<Lazy<IPackageManagerProvider, IOrderable>> packageManagerProviders)
```
Comment (emgarten) [inline src/NuGet.Clients/VsExtension/VisualStudioUIContextFactory.cs]: This should be static once _maxPackageManager is a const.

File: src/NuGet.Clients/VsExtension/VisualStudioUIContextFactory.cs [hunk 7]
```diff
@@ -22,6 +24,9 @@ internal class VisualStudioUIContextFactory : INuGetUIContextFactory
         private readonly IOptionsPageActivator _optionsPage;
         private readonly ISettings _settings;
         private readonly IDeleteOnRestartManager _deleteOnRestartManager;
+        private readonly IList<IPackageManagerProvider> _packageManagerProviders;
```
Comment (emgarten) [inline src/NuGet.Clients/VsExtension/VisualStudioUIContextFactory.cs]: This can be a normal List

File: src/NuGet.Clients/VsExtension/VisualStudioUIContextFactory.cs [hunk 8]
```diff
@@ -62,7 +69,25 @@ public INuGetUIContext Create(NuGetPackage package, IEnumerable<NuGetProject> pr
...
                 _optionsPage,
-                projects);
+                projects,
+                _packageManagerProviders);
+        }
+
+        private List<IPackageManagerProvider> Sort(IEnumerable<Lazy<IPackageManagerProvider, IOrderable>> packageManagerProviders)
+        {
+            var sortedProviders = new List<IPackageManagerProvider>();
+            var uniqueId = new HashSet<string>();
```
Comment (emgarten) [inline src/NuGet.Clients/VsExtension/VisualStudioUIContextFactory.cs]: Make this case insensitive

**Comment by emgarten** [comment]
:watch:

File: src/NuGet.Clients/PackageManagement.UI/UserInterfaceService/NuGetUIContext.cs (hunk 4 above)
Comment (zhili1208) [inline src/NuGet.Clients/PackageManagement.UI/UserInterfaceService/NuGetUIContext.cs]: sent the wrong version, updated to the right one

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
+﻿using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.VisualStudio
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: Unrelated, but please change the project name and folder from VisualStudioAPI to `NuGet.VisualStudio` the project name should match the generated assembly

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+{
+    /// <summary>
+    /// Interface for integrating other package manager.eg:bower,npm
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: Interface allowing integration of alternate package manager suggestion for a NuGet package. For example jQuery may appear on Bower and it might be more appropriate to install a package from Bower for certain projects.

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    /// Contains methods to check packages in intergated package manager and direct to intergated package manager. 
+    /// </summary>
+    [ComImport]
+    [Guid("BCED5BF2-40FC-4D9F-BF0A-43CD4E9FF65F")]
+    public interface IPackageManagerProvider
+    {
+        /// <summary>
+        /// Package manager name.
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: Is this a display name, is it localized? The Summary should clearly describe it.

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        /// </summary>
+        string PackageManagerName { get; }
+
+        /// <summary>
+        /// Package manager unique id.
+        /// </summary>
+        string PackageManagerID { get; }
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: `PackageManagerId` (note the case on d)

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+        /// <summary>
+        /// Description for package manager.
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: Again, as a consumer I don't know what to do with this, where does it gets displayed what it is for.

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+        /// </summary>
+        string Description { get; }
+
+        /// <summary>
+        /// Determine do we need to direct to integrated package manager from NuGet UI for specified package and project.
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: Check if a recommendation should be surfaced for an alternate package manager. This code should not rely on slow network calls, and should return rapidly.

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs (hunk 11 above)
Comment (zhili1208) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: Yes, it's localized. it will return something like Resources.Strings.BowerPackageProviderName from integrated Manager

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+        /// </summary>
+        /// <param name="packageId">Current package id</param>
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: NuGet package id. (dots at the end of  a sentence)

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        /// <param name="projectName">Unique project name</param>
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: What does that mean (unique?) is there a method that a user of this code can generate it?

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs (hunk 11 above)
Comment (yishaigalatzer) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: So please update the docs to reflect it

File: src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+        /// <param name="token">Cancellation Token</param>
+        /// <returns>return true if need to direct to integrated package manager for this package</returns>
+        Task<bool> CheckForPackage(string packageId, string projectName, CancellationToken token);
+
+        /// <summary>
+        /// direct to integrated package manager from NuGet UI for specified package and project.
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/VisualStudioAPI/IPackageManagerProvider.cs]: Capitalize.

This Action should take the user to the other package manager

**Comment by yishaigalatzer** [comment]
Need a test that will load multiple providers and make sure they are sorted in the right way.

:watch:

**Comment by yishaigalatzer** [comment]
Being a new NuGet API, adding the API owners
CC @pilchie @ericstj @MeniZalzman @davidfowl

File: src/NuGet.Clients/VisualStudio/IPackageManagerProvider.cs [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace NuGet.VisualStudio
+{
+    /// <summary>
+    /// Interface allowing integration of alternate package manager suggestion for a NuGet package. 
+    /// For example jQuery may appear on Bower and it might be more appropriate to install a package from Bower for certain projects 
+    /// </summary>
+    [ComImport]
+    [Guid("BCED5BF2-40FC-4D9F-BF0A-43CD4E9FF65F")]
+    public interface IPackageManagerProvider
```
Comment (emgarten) [inline src/NuGet.Clients/VisualStudio/IPackageManagerProvider.cs]: This should be IVsPackageManagerProvider

Our truly public VS interfaces have IVs on them.

File: src/NuGet.Clients/VisualStudio/IPackageManagerProvider.cs [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+        /// <param name="projectName">Unique project name for finding the project through VS dte</param>
+        /// <param name="token">Cancellation Token</param>
+        /// <returns>return true if need to direct to integrated package manager for this package</returns>
+        Task<bool> CheckForPackage(string packageId, string projectName, CancellationToken token);
+
+        /// <summary>
+        /// This Action should take the user to the other package manager.
+        /// </summary>
+        /// <param name="packageId">Current package id</param>
+        /// <param name="projectName">Unique project name for finding the project through VS dte</param>
```
Comment (emgarten) [inline src/NuGet.Clients/VisualStudio/IPackageManagerProvider.cs]: Are 3rd parties able to generate the same unique names as us? I think that logic is internal.

File: src/NuGet.Clients/VisualStudio/IPackageManagerProvider.cs [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+        /// </summary>
+        string PackageManagerName { get; }
+
+        /// <summary>
+        /// Package manager unique id.
+        /// </summary>
+        string PackageManagerId { get; }
+
+        /// <summary>
+        /// Display description for package manager when user hover over the manager name on UI.
```
Comment (emgarten) [inline src/NuGet.Clients/VisualStudio/IPackageManagerProvider.cs]: Maybe just call this the tool tip description for the package.

File: src/NuGet.Clients/PackageManagement.UI/Utility/PackageManagerProviderUtility.cs [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+namespace NuGet.PackageManagement.UI
+{
+    public static class PackageManagerProviderUtility
+    {
+        public static List<IPackageManagerProvider> Sort(IEnumerable<Lazy<IPackageManagerProvider, IOrderable>> packageManagerProviders, int max)
+        {
+            var sortedProviders = new List<IPackageManagerProvider>();
+            var uniqueId = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            foreach (var provider in Orderer.Order(packageManagerProviders))
```
Comment (emgarten) [inline src/NuGet.Clients/PackageManagement.UI/Utility/PackageManagerProviderUtility.cs:16]: :+1:

File: src/NuGet.Clients/PackageManagement.UI/Utility/PackageManagerProviderUtility.cs [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+            {
+                if (sortedProviders.Count() < max && !uniqueId.Contains(provider.Value.PackageManagerId))
```
Comment (emgarten) [inline src/NuGet.Clients/PackageManagement.UI/Utility/PackageManagerProviderUtility.cs]: This can be .Count since it is a list

File: src/NuGet.Clients/VisualStudio/IPackageManagerProvider.cs (hunk 19 above)
Comment (zhili1208) [inline src/NuGet.Clients/VisualStudio/IPackageManagerProvider.cs]: I think this is dte unique name not nuget unique name

File: NuGet.Clients.sln [hunk 23]
```diff
@@ -72,185 +72,752 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "UnitTests", "UnitTests", "{
...
+		Mono Release|x86 = Mono Release|x86
 		Release|Any CPU = Release|Any CPU
+		Release|ARM = Release|ARM
+		Release|x64 = Release|x64
+		Release|x86 = Release|x86
 	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
 		{306CDDFA-FF0B-4299-930C-9EC6C9308160}.Coverage|Any CPU.ActiveCfg = Release|Any CPU
 		{306CDDFA-FF0B-4299-930C-9EC6C9308160}.Coverage|Any CPU.Build.0 = Release|Any CPU
+		{306CDDFA-FF0B-4299-930C-9EC6C9308160}.Coverage|ARM.ActiveCfg = Debug|Any CPU
```
Comment (emgarten) [inline NuGet.Clients.sln:103]: What happened here? This doesn't look wrong, but why were these added as part of this?

Comment (zhili1208) [inline NuGet.Clients.sln:103]: I think it's because I create a new test project in clients.sln

**Comment by emgarten** [comment]
:shipit:

File: NuGet.Clients.sln [hunk 24]
```diff
@@ -72,185 +72,752 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "UnitTests", "UnitTests", "{
...
+		{1E243B39-7C61-4525-A664-3DC8589707AC}.Mono Release|x86.ActiveCfg = Release|Any CPU
+		{1E243B39-7C61-4525-A664-3DC8589707AC}.Mono Release|x86.Build.0 = Release|Any CPU
+		{1E243B39-7C61-4525-A664-3DC8589707AC}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{1E243B39-7C61-4525-A664-3DC8589707AC}.Release|Any CPU.Build.0 = Release|Any CPU
+		{1E243B39-7C61-4525-A664-3DC8589707AC}.Release|ARM.ActiveCfg = Release|Any CPU
+		{1E243B39-7C61-4525-A664-3DC8589707AC}.Release|ARM.Build.0 = Release|Any CPU
+		{1E243B39-7C61-4525-A664-3DC8589707AC}.Release|x64.ActiveCfg = Release|Any CPU
+		{1E243B39-7C61-4525-A664-3DC8589707AC}.Release|x64.Build.0 = Release|Any CPU
+		{1E243B39-7C61-4525-A664-3DC8589707AC}.Release|x86.ActiveCfg = Release|Any CPU
+		{1E243B39-7C61-4525-A664-3DC8589707AC}.Release|x86.Build.0 = Release|Any CPU
```
Comment (deepakaravindr) [inline NuGet.Clients.sln:820]: Please double-check these additions. I don't think they are valid

File: src/NuGet.Clients/PackageManagement.UI/UserInterfaceService/NuGetUIContext.cs [hunk 25]
```diff
@@ -58,5 +62,10 @@ public IEnumerable<NuGetProject> Projects
         public abstract void PersistSettings();
 
         public abstract void ApplyShowPreviewSetting(bool show);
+
+        public IEnumerable<IVsPackageManagerProvider> PackageManagerProviders
```
Comment (deepakaravindr) [inline src/NuGet.Clients/PackageManagement.UI/UserInterfaceService/NuGetUIContext.cs:66]: Please change this to be a read-only getter like SourceProvider

File: src/NuGet.Clients/PackageManagement.UI/Utility/PackageManagerProviderUtility.cs [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.VisualStudio.Utilities;
+using NuGet.VisualStudio;
+
+namespace NuGet.PackageManagement.UI
+{
+    public static class PackageManagerProviderUtility
+    {
+        public static List<IVsPackageManagerProvider> Sort(IEnumerable<Lazy<IVsPackageManagerProvider, IOrderable>> packageManagerProviders, int max)
```
Comment (deepakaravindr) [inline src/NuGet.Clients/PackageManagement.UI/Utility/PackageManagerProviderUtility.cs:11]: Break long lines. Should be <= 120 chars

File: src/NuGet.Clients/StandaloneUI/StandaloneUIContext.cs [hunk 27]
```diff
@@ -27,7 +27,7 @@ internal class StandaloneUIContext : NuGetUIContextBase
             IOptionsPageActivator optionsPageActivator,
             IEnumerable<NuGetProject> projects)
             :
-                base(sourceProvider, solutionManager, packageManager, uiActionEngine, packageRestoreManager, optionsPageActivator, projects)
+                base(sourceProvider, solutionManager, packageManager, uiActionEngine, packageRestoreManager, optionsPageActivator, projects, null)
```
Comment (deepakaravindr) [inline src/NuGet.Clients/StandaloneUI/StandaloneUIContext.cs:30]: 1) Use parameter name when parameter value is null, true or false.
2) Break long line

File: src/NuGet.Clients/VisualStudio/IVsPackageManagerProvider.cs [hunk 28]
```diff
@@ -0,0 +1,47 @@
+﻿using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.VisualStudio
+{
+    /// <summary>
+    /// Interface allowing integration of alternate package manager suggestion for a NuGet package. 
+    /// For example jQuery may appear on Bower and it might be more appropriate to install a package from Bower for certain projects
```
Comment (deepakaravindr) [inline src/NuGet.Clients/VisualStudio/IVsPackageManagerProvider.cs:9]: Please break long line here and in other places in your change as applicable

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/PackageManagerProviderTest.cs [hunk 29]
```diff
@@ -0,0 +1,66 @@
+﻿using System;
```
Comment (deepakaravindr) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/PackageManagerProviderTest.cs:1]: Remove and sort usings. Here and in other places in the PR

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/PackageManagerProviderTest.cs [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+            var sorted = PackageManagerProviderUtility.Sort(PackageManagerProviders, 3);
+
+            // Assert
+            Assert.Equal("test-version0", sorted[0].PackageManagerId);
+            Assert.Equal("test-version1", sorted[1].PackageManagerId);
+            Assert.Equal("test-version2", sorted[2].PackageManagerId);
+        }
+
+        [Fact]
+        public void PackageManagerProvider_duplicatedManagerId()
```
Comment (deepakaravindr) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/PackageManagerProviderTest.cs:36]: Should be _D? Please change it other places as appropriate

File: src/NuGet.Clients/VisualStudio/IVsPackageManagerProvider.cs [hunk 31]
```diff
@@ -0,0 +1,47 @@
...
+
+        /// <summary>
+        /// Check if a recommendation should be surfaced for an alternate package manager. 
+        /// This code should not rely on slow network calls, and should return rapidly.
+        /// </summary>
+        /// <param name="packageId">Current package id</param>
+        /// <param name="projectName">Unique project name for finding the project through VS dte</param>
+        /// <param name="token">Cancellation Token</param>
+        /// <returns>return true if need to direct to integrated package manager for this package</returns>
+        Task<bool> CheckForPackage(string packageId, string projectName, CancellationToken token);
```
Comment (deepakaravindr) [inline src/NuGet.Clients/VisualStudio/IVsPackageManagerProvider.cs:38]: It is a convention to have a suffix 'async' when the return type is Task or Task<T>. Consider changing this to 'CheckForPackageAsync' if possible.

**Comment by deepakaravindr** [comment]
:shipit: (Signed-off with comments)
